### ISSUE TITLE
morphing dice retain the "twin" property when they move from twin to non-twin size

### ISSUE THREAD

**Issue by cgolubi1** [buttonmen-dev/buttonmen#2795]
Here's a fun game that replay testing crashed on (i don't have a trace handy, because it crashed rather than bug-error-ing, so it didn't record its data, but i bet i can get one):

<img width="1149" alt="grym_vs_blackomega" src="https://user-images.githubusercontent.com/2898207/147782205-6a9f33f9-d6d5-4f4d-a502-6734b9053fb0.png">

The end of the crash trace was:

```python
  File "lib/random_ai.py", line 1003, in _is_valid_attack_of_type_Trip
    defender_min = self._min_trip_value(defender)
  File "lib/random_ai.py", line 1035, in _min_trip_value
    return min_die_value(die)
  File "lib/random_ai.py", line 193, in min_die_value
    for s in die['subdieArray']])
KeyError: 'subdieArray'
```

and the code that threw that is:
```python
   189  def min_die_value(die):
   190    if 'Twin' in die['properties']:
   191      return sum([
   192        0 if s['sides'] is 0 else 1 \
   193        for s in die['subdieArray']])
   194    return 0 if die['sides'] is 0 else 1
```

In other words, RandomAI crashed because it assumed that any time `"Twin"` appears in die properties, the die will have `"subdieArray"` defined, and that wasn't true in this case.

And, indeed, i can validate using the python API against this database state that that's the case:

```python
>>> retval.data['playerDataArray'][0]['activeDieArray'][0]
{u'description': u'Trip Morphing 4-sided die', u'skills': [u'Trip', u'Morphing'], u'recipe': u'tm(4)', u'value': 1, u'properties': [u'Twin', u'JustPerformedTripAttack', u'HasJustMorphed'], u'sides': 4}
```

**Comment by cgolubi1** [comment]
For what it's worth, this doesn't seem to necessarily make the game non-playable.  Playing via the UI, i was able to capture this slightly-malformed morphing die using a trip attack.  I imagine there would be situations in which it would cause trouble, and it's obviously incorrect, but prod site players might not notice the issue depending on what else was going on.

**Comment by blackshadowshade** [comment]
Ah, our good friend BlackOmega, generator of headaches galore.

Would you be able to grab me a responder test for this or something similar please? It should be pretty easy to fix.

**Comment by cgolubi1** [comment]
Will do!

**Comment by cgolubi1** [comment]
Ah, i bet the fact that the twin die is a swing is important --- i tried to reproduce this bug as part of responder test 57, and couldn't.

I do want to add that extra check+move to test 57, though, since i bothered to write it out and it might be useful for something someday:

```diff
diff --git a/test/src/api/responder03Test.php b/test/src/api/responder03Test.php
index 43235943..b2b24e3b 100644
--- a/test/src/api/responder03Test.php
+++ b/test/src/api/responder03Test.php
@@ -3783,6 +3783,50 @@ class responder03Test extends responderTestFramework {
             'responder004 performed Power attack using [JIMmo(S=15):15] against [%t(1,3):2]; Defender %t(1,3) was captured; Attacker JIMmo(S=15) showing 15 changed to IMmo(1,3), which then split into: IMmo(1,1) showing 2, and IMmo(0,2) showing 2. responder004 gets another turn because a Jolt die was part of the attack. ',
             $retval, array(array(1, 3), array(0, 0)),
             $gameId, 1, 'Power', 1, 0, '', array());
+        $_SESSION = $this->mock_test_user_login('responder003');
+
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder004', 'message' => 'responder004 performed Power attack using [JIMmo(S=15):15] against [%t(1,3):2]; Defender %t(1,3) was captured; Attacker JIMmo(S=15) showing 15 changed to IMmo(1,3), which then split into: IMmo(1,1) showing 2, and IMmo(0,2) showing 2'));
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder004', 'message' => 'responder004 gets another turn because a Jolt die was part of the attack'));
+        $expData['gameActionLogCount'] += 2;
+        $expData['playerDataArray'][1]['activeDieArray'][3]['value'] = 2;
+        $expData['playerDataArray'][1]['activeDieArray'][3]['sides'] = 2;
+        $expData['playerDataArray'][1]['activeDieArray'][3]['skills'] = array('Insult', 'Maximum', 'Morphing', 'Ornery');
+        $expData['playerDataArray'][1]['activeDieArray'][3]['properties'] = array('HasJustMorphed', 'HasJustSplit', 'Twin');
+        $expData['playerDataArray'][1]['activeDieArray'][3]['recipe'] = 'IMmo(1,1)';
+        $expData['playerDataArray'][1]['activeDieArray'][3]['description'] = 'Insult Maximum Morphing Ornery Twin Die (both with 1 side)';
+        $expData['playerDataArray'][1]['activeDieArray'][3]['subdieArray'] = array(array('sides' => 1, 'value' => 1), array('sides' => 1, 'value' => 1));
+        array_splice($expData['playerDataArray'][1]['activeDieArray'], 4, 0, array(array(
+            'value' => 2,
+            'sides' => 2,
+            'skills' => array('Insult', 'Maximum', 'Morphing', 'Ornery'),
+            'properties' => array('HasJustMorphed', 'HasJustSplit', 'Twin'),
+            'recipe' => 'IMmo(0,2)',
+            'description' => 'Insult Maximum Morphing Ornery Twin Die (with 0 and 2 sides)',
+            'subdieArray' => array(array('sides' => 0, 'value' => 0), array('sides' => 2, 'value' => 2)),
+        )));
+        array_splice($expData['playerDataArray'][0]['activeDieArray'], 0, 1);
+        $expData['playerDataArray'][1]['capturedDieArray'][0] = array(
+            'value' => 2,
+            'sides' => 4,
+            'recipe' => '%t(1,3)',
+            'properties' => array('WasJustCaptured', 'Twin'),
+            'subdieArray' => array(array('sides' => 1, 'value' => 1), array('sides' => 3, 'value' => 1)),
+        );
+        $expData['playerDataArray'][0]['turboSizeArray'] = array(3 => array(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30));
+        $expData['playerDataArray'][0]['capturedDieArray'][0]['properties'] = array();
+        $expData['playerDataArray'][0]['roundScore'] = 14;
+        $expData['playerDataArray'][0]['sideScore'] = -2.7;
+        $expData['playerDataArray'][1]['roundScore'] = 18;
+        $expData['playerDataArray'][1]['sideScore'] = 2.7;
+        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+
+        $_SESSION = $this->mock_test_user_login('responder004');
+        $this->verify_api_submitTurn(
+            array(),
+            'responder004 performed Power attack using [IMmo(1,1):2] against [D(4):2]; Defender D(4) was captured; Attacker IMmo(1,1) changed size from 2 to 4 sides, recipe changed from IMmo(1,1) to IMmo(4), rerolled 2 => 4. responder004\'s idle ornery dice rerolled at end of turn: IMmo(0,2) rerolled 2 => 2. ',
+            $retval, array(array(1, 3), array(0, 0)),
+            $gameId, 1, 'Power', 1, 0, '', array());
+        $_SESSION = $this->mock_test_user_login('responder003');
     }
 
     /**
```
